Source/WebKit/Shared/TextExtractionToStringConversion.cpp

```diff
@@ -0,0 +1,411 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "TextExtractionToStringConversion.h"
+
+#include <WebCore/TextExtractionTypes.h>
+#include <wtf/text/MakeString.h>
+#include <wtf/text/StringBuilder.h>
+
+namespace WebKit {
+
+using namespace WebCore;
+
+struct TextExtractionLine {
+    unsigned lineIndex { 0 };
+    unsigned indentLevel { 0 };
+};
+
+class TextExtractionAggregator : public RefCounted<TextExtractionAggregator> {
+    WTF_MAKE_NONCOPYABLE(TextExtractionAggregator);
+public:
+    TextExtractionAggregator(CompletionHandler<void(String&&)>&& completion)
+        : m_completion(WTFMove(completion))
+    {
+    }
+
+    ~TextExtractionAggregator()
+    {
+        if (auto completion = std::exchange(m_completion, { }))
```

Comment:
You don't need to do the exchange, right? Just call the handler if it hasn't been already?
```suggestion
        if (m_completion)
            m_completion(...);
```

Reply:
Ah, I generally do this to ensure there's no possibility of reentrancy (though in this case, that wouldn't happen anyways). I'll just remove the exchange.